### PR TITLE
update call map to allow multiple input parameters

### LINKED ISSUE #128072: Map with multiple arguments not supported in Dynamo and causes graph breaks
### 🐛 Describe the bug

```
import torch
torch.set_default_device('cuda')

def f(a, b):
    return (a[0] * b[0] + a[1] * b[1])

x = [torch.randn(5) for _ in range(3)]
y = [torch.randn(5) for _ in range(3)]
torch.compile(lambda x, y: map(f, x, y), fullgraph=True)(x, y)
```

```
torch._dynamo.exc.Unsupported: invalid handler args <bound method BuiltinVariable.call_map of BuiltinVariable()> [UserFunctionVariable(), ListVariable(length=3), ListVariable(length=3)] {}
```

### Versions

N/A

cc @ezyang @gchanan @zou3519 @kadeng @msaroufim @anijain2305 @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @amjames @bdhirsh

### PR DESCRIPTION
**Pull request by Aashishthakur10**
Fixes #128072 

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang

### CODE REVIEW

**Comment by bdhirsh** [comment]
Tentatively tagging @anijain2305

File: torch/_dynamo/variables/builtin.py [hunk 1]
```diff
@@ -1459,10 +1459,21 @@ def call_hasattr(self, tx, obj, attr):
...
+            if not all(len(curr) == len(unpacked[0]) for curr in unpacked):
+                raise UserError(
+                    ValueError,
+                    "map() has atleast one argument of len differing from others",
+                )
+            items = []
+            for i in range(len(unpacked[0])):
+                params = []
+                for j in range(len(unpacked)):
+                    params.append(unpacked[j][i])
```
Comment (Skylion007) [inline torch/_dynamo/variables/builtin.py:1475]:
```suggestion
                params = list(unpacked[j])
```

Comment (Aashishthakur10) [reply torch/_dynamo/variables/builtin.py:1475]: A couple of doubts here, 
1. What is the value of `j`, if we remove the inner loop?
2. Each list in `unpacked` contains at least one value for each required parameter. For example, if we have two parameters, x and y, `unpacked` includes all the variable trackers for both. unpacked[j] would give all the values for parameter `j`, instead we need to use 'unpacked[j][0]' for x and 'unpacked[j][1]' for y. So, how'd `list(unpacked[j])` work? 

An example, 
```
unpacked[0]= [LazyVariableTracker(), LazyVariableTracker(), LazyVariableTracker()] <- x values
unpacked[1]= [LazyVariableTracker(), LazyVariableTracker(), LazyVariableTracker()] <- y values
We need one of each at a time.
```

**Comment by Aashishthakur10** [comment]
Any updates on the review @Skylion007 @anijain2305 ?

File: torch/_dynamo/variables/builtin.py (hunk 1 above)
Comment (williamwen42) [reply torch/_dynamo/variables/builtin.py:1475]: I don't think this is the right suggestion.

File: torch/_dynamo/variables/builtin.py [hunk 2]
```diff
@@ -1459,10 +1459,21 @@ def call_hasattr(self, tx, obj, attr):
...
+   def call_map(self, tx, fn, *seqs):
+        if all(seq.has_unpack_var_sequence(tx) for seq in seqs):
+            unpacked = [seq.unpack_var_sequence(tx) for seq in seqs]
+
+            if not all(len(curr) == len(unpacked[0]) for curr in unpacked):
+                raise UserError(
+                    ValueError,
+                    "map() has atleast one argument of len differing from others",
+                )
+            items = []
```
Comment (williamwen42) [inline torch/_dynamo/variables/builtin.py:1471]:
```suggestion
            items = [fn.call_function(tx, args, {}) for args in zip(*unpacked)]
```

File: torch/_dynamo/variables/builtin.py [hunk 3]
```diff
@@ -1459,10 +1459,21 @@ def call_hasattr(self, tx, obj, attr):
...
 
-    def call_map(self, tx, fn, seq):
-        if seq.has_unpack_var_sequence(tx):
-            items = [fn.call_function(tx, [x], {}) for x in seq.unpack_var_sequence(tx)]
-            return variables.TupleVariable(items)
+   def call_map(self, tx, fn, *seqs):
+        if all(seq.has_unpack_var_sequence(tx) for seq in seqs):
+            unpacked = [seq.unpack_var_sequence(tx) for seq in seqs]
+
+            if not all(len(curr) == len(unpacked[0]) for curr in unpacked):
```
Comment (williamwen42) [inline torch/_dynamo/variables/builtin.py:1466]: `map` doesn't have this restriction, so we also shouldn't.

File: torch/_dynamo/variables/builtin.py [hunk 4]
```diff
@@ -1459,10 +1459,21 @@ def call_hasattr(self, tx, obj, attr):
...
+                raise UserError(
+                    ValueError,
+                    "map() has atleast one argument of len differing from others",
+                )
+            items = []
+            for i in range(len(unpacked[0])):
+                params = []
+                for j in range(len(unpacked)):
+                    params.append(unpacked[j][i])
+                items.append(fn.call_function(tx, params, {}))
```
Comment (williamwen42) [inline torch/_dynamo/variables/builtin.py:1476]: Need to return `items`

**Review by williamwen42 — CHANGES_REQUESTED** [review]
Also add the original repro to `tests/dynamo` (`test_repros.py` is probably fine). The current state of the PR doesn't work on the repro.

File: torch/_dynamo/variables/builtin.py (hunk 1 above)
Comment (Aashishthakur10) [reply torch/_dynamo/variables/builtin.py:1475]: Thanks for confirming

**Comment by williamwen42** [comment]
Commandeered in https://github.com/pytorch/pytorch/pull/130748

**Pull request closed without merging**